src/ott/solvers/linear/univariate.py

```diff
@@ -318,29 +331,120 @@ def quantile_distance(
+
+  return cost, paired_indices, mass_paired_indices
+
+
+# TODO(michalk8): better naming
+def north_west_distance(
+    x: jnp.ndarray,
+    y: jnp.ndarray,
+    cost_fn: costs.TICost,
+    a: jnp.ndarray,
+    b: jnp.ndarray,
+) -> Tuple[float, jnp.ndarray, jnp.ndarray, jnp.ndarray, jnp.ndarray]:
+  r"""Computes Univariate Distance between 1D point clouds.
 
-  i_in_sorted_x_of_quantile = all_values_sorter[i_x_cdf_inv] % n
-  i_in_sorted_y_of_quantile = all_values_sorter[i_y_cdf_inv] - n
+  Args:
+    x: TODO.
+    y: TODO.
+    cost_fn: Cost function.
+    a: TODO.
+    b: TODO.
 
-  orig_i = i_x[i_in_sorted_x_of_quantile][1:]
-  orig_j = i_y[i_in_sorted_y_of_quantile][1:]
+  Returns:
+    TODO.
+  """
 
-  return cost, jnp.stack([orig_i, orig_j]), diff_q
+  class State(NamedTuple):
+    paired_indices: jnp.ndarray
+    mass_paired_indices: jnp.ndarray
+    dual_a: jnp.ndarray
+    dual_b: jnp.ndarray
+    a: jnp.ndarray
+    b: jnp.ndarray
+
+    def replace(self, **kwargs: Any) -> "State":
+      return self._replace(**kwargs)
+
+  def dual_a_update(state: State, i: int,
+                    j: int) -> Tuple[State, jnp.ndarray, jnp.ndarray]:
+    next_ixs = jnp.array([i + 1, j])
+    val = cost_fn(x[i + 1, None], y[j, None]) - state.dual_b[j]
+    da = state.dual_a.at[i + 1].set(val)
+    return state.replace(dual_a=da), state.a[i], next_ixs
+
+  def dual_b_update(state: State, i: int,
+                    j: int) -> Tuple[State, jnp.ndarray, jnp.ndarray]:
+    next_ixs = jnp.array([i, j + 1])
+    val = cost_fn(x[i, None], y[j + 1, None]) - state.dual_a[i]
+    db = state.dual_b.at[j + 1].set(val)
+    return state.replace(dual_b=db), state.b[j], next_ixs
+
+  def body_fun(ix: int, state: State) -> State:
+    i, j = state.paired_indices[0, ix], state.paired_indices[1, ix]
+    state, min_ab, next_ixs = jax.lax.cond(
+        state.a[i] < state.b[j], dual_a_update, dual_b_update, state, i, j
+    )
+
+    pi = state.paired_indices.at[:, ix + 1].set(next_ixs)
+    mpi = state.mass_paired_indices.at[ix].set(min_ab)
+    a = state.a.at[i].set(state.a[i] - min_ab)
+    b = state.b.at[j].set(state.b[j] - min_ab)
+
+    return state.replace(paired_indices=pi, mass_paired_indices=mpi, a=a, b=b)
+
+  n, m = a.shape[0], b.shape[0]
+  q = m + n - 1
+
+  paired_indices = jnp.zeros((2, q), dtype=int)
+  mass_paired_indices = jnp.zeros(q)
+
+  state = State(
+      paired_indices=paired_indices,
+      mass_paired_indices=mass_paired_indices,
+      a=a,
+      b=b,
+      dual_a=jnp.zeros(n),
+      dual_b=jnp.zeros(m).at[0].set(cost_fn(x[0, None], y[0, None])),
+  )
+  state = jax.lax.fori_loop(0, q - 1, body_fun, state)
+
+  ot_cost = jnp.sum(state.dual_a * a) + jnp.sum(state.dual_b * b)
+
+  p_final = jnp.maximum(state.a[-1], state.b[-1])
+  mass_paired_indices = state.mass_paired_indices.at[-1].set(p_final)
+
+  return (
+      ot_cost,
+      state.paired_indices,
+      mass_paired_indices,
+      state.dual_a,
+      state.dual_b,
+  )
```

Comment:
thanks, very cool!

Reply:
thanks for tidying up my implementation, I will delete this branch https://github.com/ott-jax/ott/tree/feature/gs_1d_solver

why did the minimum cause a slow down?

Reply:
Out of interest, why is this appearing as green and not i my prev commit on the branch? Is it squashed or copied and pasted?